### PR TITLE
Updates ABI spec with hash based ids, versions and program type.

### PR DESCRIPTION
**Pull request by esdrubal**
[#5151](https://github.com/FuelLabs/sway/issues/5151)
[#5952](https://github.com/FuelLabs/sway/issues/5952)
[#5954](https://github.com/FuelLabs/sway/issues/5954)

Adds specification for how hashes based ids are generated from a string representation of types.

Updates `json-abi-format.md` with versions, program type, and hash based ids.

### Before requesting review
- [x] I have reviewed the code myself

### After merging, notify other teams

[Add or remove entries as needed]

- [ ] [Rust SDK](https://github.com/FuelLabs/fuels-rs/)
- [ ] [Typescript SDK](https://github.com/FuelLabs/fuels-ts/)
- [ ] [Sway compiler](https://github.com/FuelLabs/sway/)
- [ ] [Platform documentation](https://github.com/FuelLabs/devrel-requests/issues/new?assignees=&labels=new+request&projects=&template=NEW-REQUEST.yml&title=%5BRequest%5D%3A+) (for out-of-organization contributors, the person merging the PR will do this)
- [ ] Someone else?

### CODE REVIEW

File: src/abi/json-abi-format.md [hunk 1]
```diff
@@ -26,51 +32,54 @@ we define the following expressions:
 
 The ABI of a contract is represented as a JSON object containing the following properties:
 
-- `"types`": an array describing all the _type declarations_ used (or transitively used) in the ABI. Each _type declaration_ is a JSON object that contains the following properties:
-  - `"typeId"`: a unique integer ID.
+- `"specVersion"`: a string representing the version pointing to this document versioning. When `specVersion` is incremented, it will usually incur an increment in `abiVersion` too.
```
Comment (arboleya) [inline src/abi/json-abi-format.md:35]: Is `specVersion` replacing the `encoding` property?

encoding

File: src/abi/json-abi-format.md [hunk 2]
```diff
@@ -26,51 +32,54 @@ we define the following expressions:
...
+- `"abiVersion"`: a string representing the version of the implementation of this document `abiVersion` may be incremented while `specVersion` is not.
```
Comment (arboleya) [inline src/abi/json-abi-format.md:36]: In which situation should one care for `abiVersion`?

File: src/abi/hash-based-ids.md [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+### Tuples
+
+ `()`      => `"()"`
+ `(T1)`    => `"({abi_str(T1)})"`
+ `(T1,T2)` => `"({abi_str(T1)}, {abi_str(T2)})"`
+etc.
+
+### Enums
+
+  `Option` enum with type parameter `T` => `"enum std::option::Option<{abi_str(T)}>"`
```
Comment (nedsalk) [inline src/abi/hash-based-ids.md:51]: Does this mean that the `--json-abi-with-callpaths` is now on by default?

File: src/abi/json-abi-format.md (hunk 2 above)
Comment (IGI-111) [reply src/abi/json-abi-format.md:36]: We could have breaking changes in the spec without breaking chances in the ABI and vice versa.

Also, a new contract may want to deliberately use an old ABI version for compatibility reasons.

I really think these should be different.

Comment (arboleya) [reply src/abi/json-abi-format.md:36]: So, if the spec is at version `3` but users deliberately chose to use version `1`:

```json
{
  "specVersion": "3",
  "abiVersion": "1"
}
```

Comment (arboleya) [reply src/abi/json-abi-format.md]: @nedsalk @hal3e Did you understand the distinction between `specVersion` and `abiVersion`?

I suspect I might still be misinterpreting it.

Comment (hal3e) [reply src/abi/json-abi-format.md]: @IGI-111 can you please give us an example of a breaking change in the spec that does not break the abi and vice versa ?

Comment (arboleya) [reply src/abi/json-abi-format.md]: > Something that would bump the abiVersion would be for instance changing how intrinsic types are encoded

In this case, what's the difference between `abiVersion` and `encoding`?

Comment (IGI-111) [reply src/abi/json-abi-format.md]: If we already have an encoding field, we don't need an abiVersion one I think.

Comment (arboleya) [reply src/abi/json-abi-format.md]: @esdrubal @IGI-111 It is a NIT, but should we use the same suffix for both?
- `specVersion`
- `encodingVersion`

Renaming `encoding` -> `encodingVersion` is pretty easy for us.

Comment (arboleya) [reply src/abi/json-abi-format.md]: Thank you.

**Review by arboleya — APPROVED** [review]

**Review by IGI-111 — APPROVED** [review]

**Pull request merged**